proto/buf/alpha/registry/v1alpha1/doc.proto

```diff
@@ -377,6 +378,55 @@ message FieldOptions {
   int32 jstype = 4;
 }
 
+// FieldLiteral represents a field and its value. It can be used to model descriptor options.
+message FieldLiteral {
+  // When is_extension is true, name is the fully-qualified name of the
+  // extension. Otherwise, it is the simple name of the field.
+  string name = 1;
+  // The unique field number associated with the field.
+  int32 tag = 2;
+  // Indicates whether this is an extension or not.
+  bool is_extension = 3;
+  // The field's value.
+  LiteralValue value = 4;
+}
+
+// MessageLiteral represents an option value for a type that is a message.
+message MessageLiteral {
+  // Map of field names to value. Extension field names will be in the form:
+  // [fully.qualified.extension.Name].
+  repeated FieldLiteral fields = 1;
+}
+
+// ArrayLiteral represents an option value for a type that is a repeated field.
+// Maps are also represented using an ArrayLiteral, where each map entry is a MessageLiteral
+// that contains a key and value field.
+message ArrayLiteral {
+  repeated LiteralValue elements = 1;
+}
+
+// LiteralValue represents the actual value of the extension.
+message LiteralValue {
+  oneof value {
+    string string_value = 2;
+    int32 int32_value = 3;
+    int64 int64_value = 4;
+    uint32 uint32_value = 5;
+    uint64 uint43_value = 6;
+    fixed32 fixed32_value = 7;
+    fixed64 fixed64_value = 8;
+    sfixed32 sfixed32_value = 9;
+    sfixed64 sfixed64_value = 10;
+    float float_value = 11;
+    double double_value = 12;
+    bool bool_value = 13;
+    string enum_value = 14;
+    bytes bytes_value = 15;
+    ArrayLiteral array_value = 16;
+    MessageLiteral message_value = 17;
```

Comment:
If I'm not mistaken, this allows us to represent repeated values like they appear in the source, retaining whether they were specified as separate fields or a joined repeated field.

To elaborate - the following two options are identical:

```proto
option (google.api.http) = {
  get: "/v1/{name=messages/*}"
  additional_bindings {
    get: "xxx"
  }
  additional_bindings {
    get: "yyy"
  }
};
```

```proto
option (google.api.http) = {
  get: "/v1/{name=messages/*}"
  additional_bindings: [
    {
      get: "xxx"
    },
    {
      get: "yyy"
    }
  ]
};
```

I think it would be confusing to users if they use one form, but the docs show a different form.

Reply:
I think it will be okay to render them uniformly instead of trying to re-create the source. And I'd suggest always doing the second form (array-literal style syntax, which will be more intuitive for most readers).

For one, I think it will look nicer if the APIs are presented consistently in documentation, instead of trying to preserve possibly idiosyncratic syntax in some sources. And two, re-creating the original source from just the descriptor (i.e. trying to re-construct formatting and such from just the source code info) is not really feasible and trying to merge actual sources with the descriptor into something that is more faithful to the original source would also be a real bear to implement.